### DISCUSSION TITLE
Add custom metadata to mutations

### DISCUSSION THREAD

**Discussion by gopeter** [urql-graphql/urql#1076] [Ideas]
It would be great if we could add custom metadata to mutations. If I want to update the cache manually, I need all the variables for the current query I want to update:

```tsx
const cacheExchange = createCacheExchange({
  updates: {
    Mutation: {
      moveSitemapPage: (mutationResult, mutationArgs, cache, info) => {
        const variables: Variables = {}
        const result = mutationResult as MoveSitemapPageMutation
        const args = mutationArgs as MoveSitemapPageMutationVariables

        // ... how to fill `variables` with all the params that are needed for this query

        cache.updateQuery({ query: GetSitemapDocument, variables }, (data) => {
          return data
        })
      },
    }
  }
})
```

An idea would be something like `moveSitemapPageMutation({ input: { id, toPage } }, { extras:  { foo: "bar" } })`. `extras` could be available in the info object:

```tsx
const cacheExchange = createCacheExchange({
  updates: {
    Mutation: {
      moveSitemapPage: (mutationResult, mutationArgs, cache, info) => {
        console.log(info.extras.foo)
      }
    }
  }
})
```

**Comment by kitten** [comment]
You can populate extra variables that aren't defined in the mutation itself. That's an escape hatch we made possible explicitly in case it's never possible to have all the data be present from the cache.

It's an escape hatch because our cache assumes that all data that is needed to process a mutation is present in the variables, locally, or in the cache. That's because these are configurations matching the server's operations closely. So the idea is that if you ever have to pass extra information in it has an effect on what the UI has to pass, at which point the question arises why that information isn't in the cache.

So be aware that you can also query the cache locally with stuff like `cache.inspectFields` https://formidable.com/open-source/urql/docs/graphcache/custom-updates/#cacheinspectfields

Or read individual entries in the cache using `cache.resolve` https://formidable.com/open-source/urql/docs/api/graphcache/#resolve

But if you want a quick solution or this isn't possible, passing extra variables in that aren't in your mutation/query definition is always possible, since those are only filtered once the query/mutation is sent to your API.

**Comment by gopeter** [comment]
Holy smokes, thanks for the fast response. Actually I tried to add some metadata to my variables, but couldn‘t access them because they were filtered out 🤔

Another use case is a drag‘n‘drop sitemap creator. Once I‘ve dropped the page to another place, I need the old record (before the mutation ran), to move the pages around in the array that I have in the cache (from old page index to the new one). 

I‘m going to try to add that to the variables again. I‘m using GraphQL codegen, but don‘t think that the generated code does filtering or things like that.

**Comment by gopeter** [comment]
Yes, additional variables are filtered out:

```tsx
moveSitemapPageMutation({ input: { id, toPage }, foo: 'bar' })

const cacheExchange = createCacheExchange({
  updates: {
    Mutation: {
      moveSitemapPage: (result, args, cache, info) => {
        console.log(args)
      }
    }
  }
})
```

results to

```tsx
{
  input: {
    id: 4,
    toPage: 6
  }
}
```

instead of 

```tsx
{
  input: {
    id: 4,
    toPage: 6
  },
  foo: "bar"
}
```

Apart from this, extending the variables object breaks the typescript definitions, but that's another issue 😂

**Reply by JoviDeCroock** [reply]
I think they are filtered out of `args` but should be available on `info.variables`.

**Reply by kitten** [reply]
Oh yes, that's the missing piece here. `args` is only for the current field's arguments, so when the variables aren't filtered we mean the full `variables` input object, which is in `info.variables` as @JoviDeCroock is saying.

Also it may be beneficial for you to also check whether you can read all the information you need to perform your mutation back from the cache.

**Reply by gopeter** [reply]
Unfortunately not 😞  `moveSitemapPageMutation({ input: { id, toPage, catalogId }, foo: 'bar' })` ⬇️

<img width="1024" alt="Bildschirmfoto 2020-10-20 um 19 35 10" src="https://user-images.githubusercontent.com/1242960/96623514-c7a55000-130b-11eb-801d-0ed61a8ed324.png">

**Reply by gopeter** [reply]
> Also it may be beneficial for you to also check whether you can read all the information you need to perform your mutation back from the cache.

Unfortunately I can't read the information from the cache. The `arguments` that I get from `inspectFields` does not contain the data I need (in this case a specific field (`page`) from the record before the mutation ran) and `const page = cache.resolve({ __typename: 'SitemapPage', id: args.input.id }, 'page')` does not return the old value but the new one, since the mutation is finished when this code runs.

So it seems that passing additional data to my variables is the only way to get the "old record" in my update function.

**Reply by kitten** [reply]
That's pretty odd! If you see this line, that's the variables you'll get: https://github.com/FormidableLabs/urql/blob/935db592cf3365e29d351d654d6bdd7d3bc1e9d1/exchanges/graphcache/src/operations/write.ts#L75

`normalizeVariables` here explicitly copies over properties that aren't defined in the operation itself 🤔 https://github.com/FormidableLabs/urql/blob/935db592cf3365e29d351d654d6bdd7d3bc1e9d1/exchanges/graphcache/src/ast/variables.ts#L69-L71

**Reply by gopeter** [reply]
Here's my test case: https://codesandbox.io/s/agitated-water-si0e5

**Reply by gopeter** [reply]
I'm trying to dig into `urql`'s source code, but it's not easy to follow for me 🙈 
Anyway, I did some `console.log` hell, and that's the output so far:

<img width="1162" alt="Bildschirmfoto 2020-10-21 um 11 31 10" src="https://user-images.githubusercontent.com/1242960/96701535-02050080-1391-11eb-8f88-97c7234cc080.png">

At first, `normalizeVariables` has all of the variables, but after the mutation did run, they are lost.

**Reply by kitten** [reply]
Oh I see the problem now! Yea, that’s the gist of it. The problem being that once the mutation is sent, for us, it must then be in a sendable state, so we filter the variables as soon as the mutation leaves the cache when it’s sent to the server. I’ll have a think about that and see whether there’s another workaround in that situation

**Reply by gopeter** [reply]
Ok, thank you very much. At the moment I'm storing my metadata in a `localStorage` item (before I fire the mutation) and read it in my cache updater function afterwards. Would be great if there would be a built-in way to do this ☺️

**Reply by kitten** [reply]
So two ideas here; either we store the originally normalised variables in a `WeakMap` mapping from the operation or filtered variables to restore the missing fields, or maybe another approach for us is to make the variables that aren't in the definition non-enumerable.

**Comment by elilabes** [comment]
@kitten hey there!

I was just wondering where this got to? I have run into the same problem where I need to pass additional variables into the mutation to update the local cache. I am unable to get the variable from the existing data or from the args/mutation response so I need to pass the variable in.

The same as above seems true where any additional variables I add (after lying to typescript) are filtered out and don't exist in the `args` or `info.variables`.

I wonder if adding an additional option in addition to the `variables` key would work.

```
someMutation({
  variables: {
    varOne: "hello"
  },
  cache: { // probably needs a better name
    varTwo: "world"
  }
})
```

The cache variable could then be added to the `info` object. Just a though, but I'd be keen to help implement any solution that makes this work :smile:
